pandas/io/formats/info.py

```diff
@@ -209,147 +229,315 @@ def info(self) -> None:
+        self._lines.append(f"Data columns (total {self.col_count} columns):")
+
+    @property
+    def header_column_widths(self):
+        return [len(col) for col in self.HEADERS]
+
+    @property
+    def body_column_widths(self):
+        return [max(len(x) for x in col) for col in self.strcols]
+
+    @property
+    def gross_column_widths(self):
+        return [
+            max(header_colwidth, body_colwidth)
+            for header_colwidth, body_colwidth in zip(
+                self.header_column_widths, self.body_column_widths
+            )
+        ]
+
+    def add_header_line(self):
+        header_line = self.SPACING.join(
             [
-                _put_str(id_head, space_num),
-                _put_str(column_head, space),
-                _put_str(count_header, space_count),
-                _put_str(dtype_header, space_dtype),
+                _put_str(header, col_width)
+                for header, col_width in zip(self.HEADERS, self.gross_column_widths)
             ]
         )
-        lines.append(header)
+        self._lines.append(header_line)
 
-        top_separator = "".join(
+    def add_separator_line(self):
+        separator_line = self.SPACING.join(
             [
-                _put_str("-" * len_id, space_num),
-                _put_str("-" * len_column, space),
-                _put_str("-" * len_count, space_count),
-                _put_str("-" * len_dtype, space_dtype),
+                _put_str("-" * header_colwidth, gross_colwidth)
+                for header_colwidth, gross_colwidth in zip(
+                    self.header_column_widths, self.gross_column_widths
+                )
             ]
         )
-        lines.append(top_separator)
-
-        for i, col in enumerate(ids):
-            dtype = dtypes[i]
-            col = pprint_thing(col)
-
-            line_no = _put_str(f" {i}", space_num)
-            count = ""
-            if show_counts:
-                count = counts[i]
-
-            lines.append(
-                line_no
-                + _put_str(col, space)
-                + _put_str(count_temp.format(count=count), space_count)
-                + _put_str(dtype, space_dtype)
+        self._lines.append(separator_line)
+
+    def add_body_lines(self):
+        strrows = list(zip(*self.strcols))
+        for row in strrows:
+            body_line = self.SPACING.join(
+                [
+                    _put_str(col, gross_colwidth)
+                    for col, gross_colwidth in zip(row, self.gross_column_widths)
+                ]
             )
-
-    def _non_verbose_repr(self, lines: List[str], ids: "Index") -> None:
-        lines.append(ids._summary(name="Columns"))
+            self._lines.append(body_line)
+
+
+class DataFrameTableBuilderVerboseNoCounts(DataFrameTableBuilderVerbose):
+    """Verbose info table builder without non-null counts column."""
+
+    HEADERS = [
+        " # ",
+        "Column",
+        "Dtype",
```

Comment:
can you remove the trailing comma so it fits on one line?

Reply:
Did that. But IMHO, it is more clear, when it is multilne. Anyway, did as you suggested.